# Analysis of PELT signal error due to skip last window update

In [1]:
import logging
from conf import LisaLogging
LisaLogging.setup()

2017-02-17 19:51:33,920 INFO    : root         : Using LISA logging configuration:
2017-02-17 19:51:33,922 INFO    : root         :   /data/Code/lisa/logging.conf


In [60]:
# Generate plots inline
%matplotlib inline
import json
import os
from trace import Trace
import numpy
import pandas as pd
import matplotlib.pyplot as plt
import trappy

path_to_dat = "/home/joelaf/repo/lisa-aosp/external/lisa/ipynb/scratchpad/pelt-error/trace.dat"

## Parse Trace and Profiling Data

In [3]:
trace = Trace(None, path_to_dat, events=[ 'sched_switch', 'pelt_update' ])

## Trace visualization

In [4]:
trappy.plotter.plot_trace(trace.ftrace)

## Latency DataFrames

In [64]:
df = trace.data_frame.trace_event('pelt_update')
rq_df = df[df.cfs_rq == 1]
rq_df.columns

Index([u'__comm', u'__cpu', u'__line', u'__pid', u'__tgid', u'acc_load_avg',
       u'acc_util_avg', u'cfs_rq', u'delta_us', u'load_avg', u'load_err',
       u'load_sum', u'sum_err', u'util_avg', u'util_err'],
      dtype='object')

In [61]:
# Plot the accurate and the actual signals for the RQ
print 'UTIL ERROR'
trappy.ILinePlot(trace.ftrace,
                signals = [
                    'pelt_update:util_avg',
                    'pelt_update:acc_util_avg',
                ]).view()

print 'LOAD ERROR'
trappy.ILinePlot(trace.ftrace,
                signals = [
                    'pelt_update:load_avg',
                    'pelt_update:acc_load_avg',
                ]).view()

UTIL ERROR


LOAD ERROR


Data showing util/load errors on the RQ (cpu 1) when thread0 running
-------------------------------
Note that, as expected, the error exists only for cases where delta is < 1ms (now - last_update_time)

In [83]:
err_df = rq_df[(rq_df.util_err > 1) | (rq_df.load_err > 1)][rq_df['__comm'] == 'thread0']
err_df = err_df[['__comm',  'acc_load_avg', 'util_avg', 'acc_util_avg', \
                 'util_err', 'delta_us', 'load_sum', 'sum_err', 'load_avg', 'load_err']]

err_df.sort(columns=['delta_us'])
print 'number of errors: ' + str(len(err_df))
err_df


number of errors: 51


/home/joelaf/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if __name__ == '__main__':
/home/joelaf/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:4: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


__comm  acc_load_avg  util_avg  acc_util_avg  util_err  delta_us  \
Time                                                                          
0.212311  thread0           108        83            87         4       194   
0.212334  thread0           108        83            88         5        26   
0.232412  thread0           411       395           398         3       106   
0.615666  thread0            81        74            79         5       236   
0.615689  thread0            81        74            80         6        27   
0.716483  thread0            80        76            79         3       122   
0.736533  thread0           401       380           398        18       854   
1.242656  thread0           403       396           402         6       276   
1.749170  thread0           403       399           402         3       121   
2.152743  thread0           403       394           401         7       330   
2.455914  thread0           396       390           394         4       159   
3.062918  thread0           398       387           397        10       450   
3.749162  thread0            88        75            87        12       519   
3.749187  thread0            88        75            87        12        28   
3.756066  thread0           210       204           209         5       241   
4.072932  thread0           404       394           403         9       392   
4.355859  thread0            85        73            83        10       435   
4.355883  thread0            85        73            83        10        27   
4.376122  thread0           405       398           404         6       263   
4.761087  thread0            80        76            78         2       116   
4.762599  thread0           111       105           106         1        13   
4.882727  thread0           403       395           399         4       190   
4.983089  thread0           398       385           396        11       545   
4.983090  thread0           398       385           396        11         1   
5.165969  thread0           107       104           106         2        67   
5.165983  thread0           108       104           106         2        17   
5.169245  thread0          1819       160           165         5       207   
5.173028  thread0          1781       222           231         9       437   
5.182618  thread0          2092       373           374         1        56   
5.286056  thread0           605       400           403         3       154   
5.366925  thread0           117        76            78         2       117   
5.485995  thread0           369       363           365         2        68   
5.672669  thread0           110       107           109         2        47   
5.772599  thread0            80        75            78         3       140   
5.772628  thread0            81        75            79         4        31   
5.772638  thread0            81        75            79         4        11   
5.776006  thread0           146       141           143         2        69   
5.792802  thread0           404       396           401         5       205   
5.982620  thread0           227       211           225        14       684   
6.075950  thread0            85        75            84         9       435   
6.075973  thread0            86        75            85        10        27   
6.086009  thread0           259       256           258         2        63   
6.277303  thread0            79        75            78         3       117   
6.582720  thread0           105       102           104         2        72   
6.582736  thread0           106       102           104         2        20   
6.986014  thread0            90        76            88        12       590   
6.986053  thread0            91        76            89        13        45   
6.996056  thread0           262       259           261         2        63   
7.309478  thread0           405       398           401      

# Summary of issues

### * At 5.98s, there is a 6% error in util_avg 225 vs 211) - this causes a glitch and makes the signal less smooth
----
![glitch](pelt-signal-glitch-error-6pc.png)


### * At 3.06s, there is a 3% error in util_avg - causing lowered peak of util_avg (397 -> 387) with delta ~450us
![lower-peak](pelt-3percent-error.png)
